File: packages/node/src/files.ts
import { File as BufferFile } from 'node:buffer';

/**
 * Node-native file input for multipart endpoints (`demands.uploadDocument`,
 * `timestamps.create`). Node servers hold file bytes as `Buffer`s (from
 * `multer`, `fs.readFile`, an upstream fetch, ...) — not browser `File`
 * objects — so the facade accepts bytes + filename and builds the
 * generated client's expected `File` itself.
 */
export interface FileInput {
  /** Raw file bytes. */
  content: Buffer | Uint8Array;
  /** Original filename, including extension (e.g. `"sozlesme.pdf"`). Required — the server infers processing (PDF vs image vs office doc) from the extension. */
  filename: string;
  /** MIME type. Best-effort inferred server-side from the extension when omitted. */
  contentType?: string;
}

/**
 * Builds a spec-compliant `File` for multipart upload from Node-native
 * bytes, using `node:buffer`'s `File` (stable since Node 18.13, and the
 * same class as the global `File`/`FormData` used by axios's Node adapter —
 * `@imzala/node` targets servers, so we don't assume a browser global).
 *
 * Cast to the generated client's (DOM-lib-sourced) `File` type: `node:buffer`'s
 * `File` is runtime-identical to `globalThis.File` but TypeScript sees them
 * as distinct declarations, so a structural cast is needed here.
 */
export function toUploadFile(input: FileInput): File {
  const init = input.contentType ? { type: input.contentType } : undefined;
  return new BufferFile([input.content], input.filename, init) as unknown as File;
}

export interface UploadPartyInput {
  first_name: string;
  last_name: string;
  /** Email or phone (or both) required per party. */
  email?: string;
  /** E.164 format (e.g. `"+905551234567"`). */
  phone?: string;
}

export interface UploadDemandParams {
  /** One document OR 1-20 images — merged server-side into a single PDF. */
  files: FileInput[];
  parties: UploadPartyInput[];
  /** Reorders a multi-image upload — indices into `files`. */
  order?: number[];
  title?: string;
  description?: string;
}

export interface CreateTimestampParams extends FileInput {
  /** Client-generated idempotency key (UUID recommended) — replays within 5 minutes return the original result without spending a credit. */
  idempotencyKey?: string;
  description?: string;
  ownerFirstName?: string;
  ownerLastName?: string;
}

File: packages/dotnet/src/Files.cs
using System.Text.Json.Serialization;
using GeneratedFileParameter = ImzalaApiClient.Client.FileParameter;

namespace ImzalaSdk;

/// <summary>
/// .NET-native file input for multipart endpoints (<c>Demands.UploadDocumentAsync</c>,
/// <c>Timestamps.CreateAsync</c>). Servers hold file bytes as <c>byte[]</c> (from
/// <c>IFormFile</c>, <c>File.ReadAllBytes</c>, an upstream download, ...) — the
/// facade accepts bytes + filename and builds the vendored client's
/// <c>FileParameter</c> itself (unlike TS, which has to cast <c>node:buffer</c>'s
/// <c>File</c> to the DOM-lib-sourced generated type — the httpclient C#
/// generator's <c>FileParameter(filename, contentType, Stream)</c> constructor is
/// already ergonomic, no cast/shim needed).
/// </summary>
public sealed class FileInput
{
    /// <summary>Raw file bytes.</summary>
    public required byte[] Content { get; init; }

    /// <summary>Original filename, including extension (e.g. <c>"sozlesme.pdf"</c>). Required — the server infers processing (PDF vs image vs office doc) from the extension.</summary>
    public required string FileName { get; init; }

    /// <summary>MIME type. Best-effort inferred server-side from the extension when omitted.</summary>
    public string? ContentType { get; init; }

    /// <summary>Builds the vendored client's <c>FileParameter</c> from this input's bytes.</summary>
    internal GeneratedFileParameter ToFileParameter() =>
        new(FileName, ContentType ?? "application/octet-stream", new MemoryStream(Content));
}

/// <summary>One signing party for <c>Demands.UploadDocumentAsync</c>. Email or phone (or both) required per party.</summary>
public sealed class UploadPartyInput
{
    [JsonPropertyName("first_name")]
    public required string FirstName { get; init; }

    [JsonPropertyName("last_name")]
    public required string LastName { get; init; }

    [JsonPropertyName("email")]
    public string? Email { get; init; }

    /// <summary>E.164 format (e.g. <c>"+905551234567"</c>).</summary>
    [JsonPropertyName("phone")]
    public string? Phone { get; init; }
}

/// <summary>Parameters for <c>Demands.UploadDocumentAsync</c> — creates a demand directly from an uploaded document (no template).</summary>
public sealed class UploadDemandParams
{
    /// <summary>One document OR 1-20 images — merged server-side into a single PDF.</summary>
    public required IReadOnlyList<FileInput> Files { get; init; }

    public required IReadOnlyList<UploadPartyInput> Parties { get; init; }

    /// <summary>Reorders a multi-image upload — indices into <see cref="Files"/>.</summary>
    public IReadOnlyList<int>? Order { get; init; }

    public string? Title { get; init; }

    public string? Description { get; init; }
}

/// <summary>Parameters for <c>Timestamps.CreateAsync</c>.</summary>
public sealed class CreateTimestampParams
{
    public required byte[] Content { get; init; }

    public required string FileName { get; init; }

    public string? ContentType { get; init; }

    /// <summary>Client-generated idempotency key (UUID recommended) — replays within 5 minutes return the original result without spending a credit.</summary>
    public string? IdempotencyKey { get; init; }

    public string? Description { get; init; }

    public string? OwnerFirstName { get; init; }

    public string? OwnerLastName { get; init; }
}
